pkg/cloud/services/s3/s3.go

```diff
@@ -0,0 +1,249 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package s3
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+	"github.com/pkg/errors"
+
+	iam "sigs.k8s.io/cluster-api-provider-aws/iam/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope"
+)
+
+// Service holds a collection of interfaces.
+// The interfaces are broken down like this to group functions together.
+// One alternative is to have a large list of functions from the ec2 client.
+type Service struct {
+	scope     scope.S3Scope
+	S3Client  s3iface.S3API
+	STSClient stsiface.STSAPI
+}
+
+// NewService returns a new service given the api clients.
+func NewService(s3Scope scope.S3Scope) *Service {
+	s3Client := scope.NewS3Client(s3Scope, s3Scope, s3Scope, s3Scope.InfraCluster())
+	STSClient := scope.NewSTSClient(s3Scope, s3Scope, s3Scope, s3Scope.InfraCluster())
+
+	return &Service{
+		scope:     s3Scope,
+		S3Client:  s3Client,
+		STSClient: STSClient,
+	}
+}
+
+func (s *Service) ReconcileBucket() error {
```

Comment:
It would be good to have some logging in here to help with debugging, if possible.

Reply:
What would you like to be logged?

I think logging creating bucket operation does not make sense, as we do not distinguish between checking for existence and actually creating the bucket, so the log message will be printed every reconciliation, which isn't good. I'm fine with changing the implementation, though for setting bucket policy it may not be as easy IIRC.

Also, as far as I remember, possible error messages are also expressive enough for basic debugging.

Reply:
I was thinking something very basic like `creating bucket......` and  `ensuring bucket policy....` at a higher verbsoity level just to help with debugging. But this is just a preference and if you think any error messages are expressive enough then ignore this :).

Reply:
At a glance I haven't seen a way for higher verbosity levels for scope logging, but maybe klog can be used directly. then indeed it make sense to add some logging. I'll have a look.

Reply:
I've added some log messages which I think make sense. I'll mark this conversation as resolved, so we can make another round of review.